core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -756,10 +757,10 @@ private[spark] class ExecutorAllocationManager(
       val taskId = taskEnd.taskInfo.taskId
       val taskIndex = taskEnd.taskInfo.index
       val stageId = taskEnd.stageId
+      val stageAttemptId = taskEnd.stageAttemptId
+      val stageAttempt = (stageId, stageAttemptId)
```

Comment:
IIRC, we receive only one `SparkListenerStageCompleted` event for multiple stage attempts. And, it's possible that `SparkListenerTaskEnd` comes from previous stage attempt after we've received `SparkListenerStageCompleted`. In that case, we need to clean up related structures for those stage attempts in `onTaskEnd`(if this is a final  finished task within corresponding stage attempt) as we already do in `onStageCompleted`.

Reply:
In the stage retry process, `StageCompleted` will be sent first and then `StageSubmitted` will be sent, so `StageCompleted` will be received multiple times.
https://github.com/apache/spark/pull/24497#discussion_r280147031

Based on this statement, the corresponding stage attempt is cleaned up when `StageCompleted` is processed, so the `TaskEnd` after `StageCompleted` should have no relevant structure.
Unless `EventBus` is full, the `StageCompleted` is discarded.

Reply:
Oh yes, I think you're right.